README.md
# Tidy Assets
##  A plugin to rearrange [Zenphoto](https://www.zenphoto.org/) resources

### What it does
Tidy Assets shifts all Zenphoto JavaScript elements, including inline scripts and optionally CSS resources, to the bottom of the _body_ element. This could improve the user experience on their first visit to your site, by delaying render-blocking resources.

It also offers the option to remove the jQuery Migrate plugin, which is included with Zenphoto since v1.6 to ensure compatibility with older themes and plugins.

Both of the above features could have a positive SEO impact, since the loading speed of a page is one of the parameters involved in determining its ranking.

Attributes other than _rel=“stylesheet”_ of moved elements are not reflected in the output. In most cases, these are attributes that are not actually needed, such as _type=“text/javascript”_ and _type=“text/css”_, unnecessary in HTML5, or _async_ and _defer_, which lose their importance after the resources have been moved to the bottom of the _body_ element.

In addition to the CSS and JavaScript elements, the remaining HTML is also cleaned up slightly, by removing blanks from the edges of lines and replacing multiple instances of white spaces or tabs with a single white space. Empty class attributes of HTML elements are dropped too.

### How it works
Tidy Assets collects the most part of the Zenphoto PHP output, starting from the first line printed by the `theme_head` registered filters and stopping at the position where the `theme_body_close` register is placed in the theme scripts, which is just before the closing body tag. To do so, the filters of this plugin are registered with a priority of 99999. The collected output is then parsed into JS, CSS and Other Items. CSS and JS file references with duplicate base names, if any, are reduced to just the first instance (paths are ignored). The inline JS is consolidated into one script tag only, and can be optionally minified “on the fly”, with [JShrink](https://github.com/tedious/JShrink) or [JSqueeze](https://github.com/tchwork/jsqueeze). At this point, Other Items are printed first, followed by CSS file references, JS file references and finally the consolidated inline script.

That said, if any plugin sets a priority greater than 99999 for its `theme_head` filter, its output will of course not get processed. The reverse happens for the `theme_body_close` filter, so the output of filters with a priority lower than 99999 (all of them, I hope) will be excluded from processing, leaving the order of its items unchanged. This allows themes and other plugins to pass variables from PHP to JavaScript if they use `theme_body_close` for their JS resources.

Elements belonging to groups to be moved, which must instead remain in their original positions for important reasons (i.e. some CSS needed at the top of head section or some analytical JS), can be kept apart from processing by using a filter option filled with suitable identification strings.

### What it takes
All of the above comes at a very low cost in terms of processing time, which with default options is **less than 1 millisecond** on both my local server and my live server. Enabling the inline JS minification, processing time rises to ∼ 2 - 15ms, depending on the number of plugins enabled, the server CPU speed/load and the algorithm chosen for minification. By removing unnecessary characters, JShrink reduces JS size by approximately 20%. It’s a bit faster than JSqueeze, which also performs some further optimizations in return saving up to 30% of the original size. In my tests, without any pretense of completeness, the difference in processing time between the two algorithms seems very similar to the difference in compression ratios.

Keep in mind, however, that this “on the fly” minification, once the inline code has already been shifted at the bottom of the DOM, is almost always useless, when it is not counterproductive, for SEO/performance optimization purposes, especially if gzip compression is enabled on the server.

***

### Installation
1. Download the latest release and extract the content
2. Use an FTP client to upload **only** the `tidy-assets` folder and the `tidy-assets.php` file into the `plugins` folder of your ZP installation
3. Activate **tidy-assets** in the Plugins > Seo tab of your admin area
4. Visit the plugin options page to set them according to your needs

### Translations
Tidy Assets is translatable by configuring a [Poedit](https://poedit.net/) catalogue with the keywords `gettext_pl`, `ngettext_pl:1,2`. If you wish to provide a translation other than Italian, which is already available, be sure to uncheck the box _Also use default keywords for supported languages_ in your catalogue properties, so that you don’t have to translate strings already managed by Zenphoto translators, but just the plugin specific strings.

### Credits
This plugin has been developed starting from [headConsolidator v1.4.3](https://www.zenphoto.org/news/headConsolidator/) by Stephen Billard (sbillard).

It includes [JShrink](https://github.com/tedious/JShrink) and [JSqueeze](https://github.com/tchwork/jsqueeze) for optional “on the fly” minification. Refer to their sites for licensing information.

Special thanks to [acrylian](https://github.com/acrylian) for suggestions and support.


CHANGELOG.md
# Changelog
## [Unreleased] - 2023-01-03

### Fixed
* Rare JavaScript errors like _(intermediate value)(...) is not a function_ were resolved by adding a semicolon after each block of internal code, before consolidating.

## [v1.1] - 2023-01-03

### Added
* Preserve formatting of `<pre>` tags in html output
* An option to remove jQuery Migrate from the site front-end

### Changed
* Implemented `preg_split` to detect end-of-line
* CDN links updated to jQuery 3.6.3 but for Google (3.6.1) and Microsoft (3.6.0)

### Removed
* The latest version of jQuery for self hosting is no longer included, as it is now provided by Zenphoto (since ZP v1.6)


[Unreleased]: https://github.com/bic-ed/Tidy-Assets/compare/v1.1..master
[v1.1]: https://github.com/bic-ed/Tidy-Assets/compare/v1.0.0...v1.1


tidy-assets.php
<?php
/**
 * Developed starting from
 * {@link https://www.zenphoto.org/news/headConsolidator/ headConsolidator v1.4.3}
 * by Stephen Billard (sbillard).
 *
 * Tidy Assets collects the most part of the Zenphoto’s PHP output, starting from the
 * first line printed by the "theme_head" registered filters and stopping at the position
 * where the "theme_body_close" filter registering is placed in the PHP theme scripts,
 * which is just before the closing body tag. The collected output is then parsed into JS,
 * CSS and Other Items. The internal JS is consolidated into one script tag only, and can
 * be optionally minified “on the fly”, with JShrink or JSqueeze. At the end, Other Items
 * are printed first, followed by CSS files references, JS files references and at last by
 * the consolidated internal script.
 *
 * @author Antonio Ranesi (bic-ed)
 * @license GPL v2 or later
 * @package plugins
 * @subpackage tidy-assets
 */

$plugin_is_filter = 1 | THEME_PLUGIN;
$plugin_description = gettext_pl("Shifts all Zenphoto JavaScript elements, including internal scripts and optionally CSS resources, to the bottom of the body element. It also offers the opportunity to remove the jQuery Migrate plugin, which is included with Zenphoto to ensure compatibility with older themes and plugins.", "tidy-assets");
$plugin_author = 'Antonio Ranesi (bic-ed)';
$plugin_version = '1.1';
$plugin_date = '03/01/2023';
$plugin_category = gettext('SEO');
$plugin_URL = "http://www.antonioranesi.it/pages/tidy-assets-zenphoto-plugin";
$plugin_siteurl = "http://www.antonioranesi.it/pages/tidy-assets-zenphoto-plugin";

$option_interface = 'tidyAssetsOptions';

zp_register_filter('theme_head', 'tidyAssets::startCollecting', 99999);
zp_register_filter('theme_body_close', 'tidyAssets::parseAndOutput', 99999);


class tidyAssetsOptions {
  /**
  * class instantiation function
  *
  */
  function __construct() {
    setOptionDefault('tidy-assets_jquery', false);
    setOptionDefault('tidy-assets_jq_migrate', 0);
    setOptionDefault('tidy-assets_css', 0);
    setOptionDefault('tidy-assets_comments', 1);
    setOptionDefault('tidy-assets_minify', false);
    setOptionDefault('tidy-assets_skip', '');
  }

  function getOptionsSupported() {
    $jshrink = '<a rel="noopener" target="_blank" href="https://github.com/tedious/JShrink">JShrink</a>';
    $jsqueeze = '<a rel="noopener" target="_blank" href="https://github.com/tchwork/jsqueeze">JSqueeze</a>';
    $options = array(
      gettext_pl("jQuery Host", "tidy-assets") => array(
        'key' => 'tidy-assets_jquery',
        'type' => OPTION_TYPE_SELECTOR,
        'order' => 90,
        'selections' => array(
          // '1.' . gettext_pl("Self hosted", "tidy-assets") => '/plugins/tidy-assets/jquery/jquery-3.5.1.min.js',
          'jQuery' => 'https://code.jquery.com/jquery-3.6.3.min.js" integrity="sha256-pvPw+upLPUjgMXY0G+8O0xUf+/Im1MZjXxxgOcBQBXU=" crossorigin="anonymous',
          'Google' => 'https://ajax.googleapis.com/ajax/libs/jquery/3.6.1/jquery.min.js',
          'Microsoft' => 'https://ajax.aspnetcdn.com/ajax/jQuery/jquery-3.6.0.min.js',
          'Cloudflare' => 'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.3/jquery.min.js',
          'jsDelivr' => 'https://cdn.jsdelivr.net/npm/jquery@3.6.3/dist/jquery.min.js',
        ),
        'null_selection' => gettext_pl("Self Hosted", "tidy-assets"),
        'desc' => gettext_pl("Select your preferred server among self hosting (recommended) or one of the CDN services listed on the jQuery website.", "tidy-assets") . " "
        . gettext_pl("The self hosted version is the one provided by your Zenphoto installation.", "tidy-assets")
        . " <strong>"
        . gettext_pl("The CDNs hosted version is", "tidy-assets")
        . " v3.6.3, but the one from Google (v3.6.1) and Microsoft (v3.6.0) </strong></p>"
        . "<p class='notebox'>"
        . gettext_pl("<strong>Notice:</strong> Provision of assets from a CDN service is explicitly discouraged by Zenphoto guidelines. It may result in the need to be accounted for in your privacy policy terms (in some jurisdictions). Select <strong>Self Hosted</strong> if you are not sure how to do it correctly or if you want to keep full control of your site.", "tidy-assets")
        . "</p>"
      ),
      gettext_pl("Remove jQuery Migrate", "tidy-assets") => array(
        'key' => 'tidy-assets_jq_migrate',
        'type' => OPTION_TYPE_CHECKBOX,
        'order' => 15,
        'desc' => gettext_pl("Check to remove jQuery Migrate from your site front-end.", "tidy-assets")
        . "<br>"
        . "<p class='notebox'>" . gettext_pl("Check that everything works as expected after enabling this option.", "tidy-assets")
      ),
      gettext_pl("Apply to CSS assets", "tidy-assets") => array(
        'key' => 'tidy-assets_css',
        'type' => OPTION_TYPE_CHECKBOX,
        'order' => 20,
        'desc' => gettext_pl("Shifts the CSS assets to the bottom of the <code>body</code> element as well, just before the JavaScript elements.", "tidy-assets")
      ),
      gettext_pl("Output comments", "tidy-assets") => array(
        'key' => 'tidy-assets_comments',
        'type' => OPTION_TYPE_CHECKBOX,
        'order' => 30,
        'desc' => gettext_pl("Print comments in HTML output to let you know what has been relocated and where. Processing time is printed as well, in a comment at the end of the plugin output.", "tidy-assets")
      ),
      gettext_pl("Minify internal JavaScript", "tidy-assets") => array(
        'key' => 'tidy-assets_minify',
        'type' => OPTION_TYPE_SELECTOR,
        'order' => 40,
        'selections' => array(
          gettext_pl("With JSqueeze", "tidy-assets") => 'tidy-assets_jsqueeze',
          gettext_pl("With JShrink", "tidy-assets") => 'tidy-assets_jshrink',
        ),
        'null_selection' => gettext('No'),
        'desc' => sprintf(gettext_pl("By removing unnecessary characters, %s reduces the consolidated script size by approximately 20%%. It’s a bit faster than %s, which in return does some further optimizations as well, sparing up to 30%% of the original size. The difference in processing time between the two algorithms is very similar to the difference in compression ratios. Note that this <strong>“on the fly” minification</strong>, once internal code has already been shifted at the bottom of the DOM, <strong>is almost always useless</strong> when not counterproductive for SEO/performance optimization purposes, especially if gzip compression is enabled on the server.", "tidy-assets"), $jshrink, $jsqueeze)
      ),
      gettext_pl("Skip elements", "tidy-assets") => array(
        'key' => 'tidy-assets_skip',
        'type' => OPTION_TYPE_TEXTAREA,
        'order' => 45,
        'multilingual' => false,
        'desc' => '<p>'
        . gettext_pl("This is a filter used to prevent shifting specific elements that must remain in their original positions. For each of these items, if you have any, enter an unique identification string here, one per line. Leave blank to disable filtering.", "tidy-assets")
        . '</p><p>'
        . gettext_pl("For JS and CSS files, the best candidate is of course the file link, without quotes. If the link for a CSS file is <code>href=\"some_path/some_file.css\"</code>, you should use <code>some_path/some_file.css</code>, but even just <code>some_file.css</code> might be enough.", "tidy-assets")
        . '</p><p>'
        . gettext_pl("For an internal JS script, you will need to spot some part of the code that is not repeated identically in any other script. It could be the name of a function, your analytics code or so. A script containing <code>function pageselectCallback</code> could be added to the list by typing <code>function pageselectCallback</code> or even just <code>pageselectCallback</code>, provided that there are no other internal script containing the same string.", "tidy-assets")
        . '<p>'
      ),
    );
    return $options;
  }
  function handleOption($option, $currentValue) {

  }
}

class tidyAssets {

  static function startCollecting() {
    ob_start();
  }

  static function parseAndOutput() {
    $data = ob_get_contents();
    ob_end_clean();
    /**
     * Unique identification strings to recognize the elements not to be moved
     * @var array
     */
    $skip = preg_split('~\R~', getOption('tidy-assets_skip'));
    foreach ($skip as $key => $value) {
      if (empty($value)) {
        unset($skip[$key]);
      }
    }

    if (getOption('tidy-assets_comments')) {
      $start = microtime(true);
    }

    // JavaScript assets
    $matches = tidyAssets::extract($data, '~<script.*src="(.*)".*></script>~mU', $skip);
    $js_files = array();
    $use_cdn_jq = getOption('tidy-assets_jquery');
    $remove_jq_migrate = getOption('tidy-assets_jq_migrate');
    while (!empty($matches[1])) { // flush out the duplicates. Earliest wins
      $file = array_pop($matches[1]);
      if ($use_cdn_jq && $file == "/zp-core/js/jquery.min.js") { // Use CDN for jQuery
        $file = $use_cdn_jq;
      }
      if ($remove_jq_migrate && $file == "/zp-core/js/jquery-migrate.min.js") { // Remove jQuery Migrate
        continue;
      }
      $js_files[basename($file)] = $file;
    }

    // Internal JavaScript
    $matches = tidyAssets::extract($data, '~<script.*>(.*)</script>~mUs', $skip);
    $internal_js = $matches[1];
    if (!empty($internal_js)) {
      $internal_js_output = '';
      foreach ($internal_js as $some_js) {
        $some_js = trim($some_js);
        // Ensure a semicolon after each internal js.
        $semicolon = '';
        if (substr($some_js, -1) != ';') {
          $semicolon = ';';
        }
        $internal_js_output .= $some_js . $semicolon . "\n";
      }
    }

    // CSS assets (if requested)
    if (getOption('tidy-assets_css')) {
      $matches = tidyAssets::extract($data, '~<link.*rel="stylesheet".*href="(.*)".*>~mU', $skip);
      $css_files = array();
      while (!empty($matches[1])) { // flush out the duplicates. Earliest wins
        $file = array_pop($matches[1]);
        $css_files[basename($file)] = $file;
      }
    }

    // Other Items
    // While we are here, let's do some additional cleaning (if not in a <pre> tag)
    $other_items = preg_split('~\R~', $data);
    $in_a_pre_tag = false;
    foreach ($other_items as $key=>$line) {
      if ($in_a_pre_tag === false && strpos($line, '<pre') !== false) {
        if (strpos($line, '</pre>') === false) {
          $in_a_pre_tag = true;
        }
        $line = ltrim($line);
        $other_items[$key] = $line;
        continue;
      }
      if ($in_a_pre_tag === true && strpos($line, '</pre>') === false) {
        continue;
      } else if (strpos($line, '</pre>') !== false) {
        $in_a_pre_tag = false;
        $line = rtrim($line);
        $other_items[$key] = $line;
        continue;
      }
      // If we got here, we are not in a <pre> tag
      $in_a_pre_tag = false;
      $line = trim($line);
      if (empty($line)) {
        unset($other_items[$key]);
      } else {
        // 1.Removing the empty class attribute produced sometimes by some Zenphoto functions
        $line = str_replace(' class=""', '', $line);
        // 2.Replacing multiple spaces/tabs with a single space
        $line = preg_replace('/\s\s+/', ' ', $line);
        $other_items[$key] = $line;
      }
    }

    // Start printing the output: Other Items...
    if (isset($start)) {
      echo "<!-- Tidy Assets Start-> Unmoved items, no changes before this line -->\n";
    }
    $other_items = implode("\n", $other_items);
    echo $other_items . "\n";

    // ...CSS Assets...
    if (!empty($css_files)) {
      $css_files = array_reverse($css_files);
      $css_files = '<link rel="stylesheet" href="'
      . implode('">' . "\n" . '<link rel="stylesheet" href="', $css_files) . '">' . "\n";
      if (isset($start)) {
        echo "<!-- Tidy Assets-> CSS files -->\n";
      }
      echo $css_files;
    }

    // ...JavaScript Assets...
    if (!empty($js_files)) {
      $js_files = array_reverse($js_files);
      $js_files = '<script src="'
      . implode('"></script>'."\n".'<script src="', $js_files) . '"></script>'."\n";
      if (isset($start)) {
        echo "<!-- Tidy Assets-> JavaScript files -->\n";
      }
      echo $js_files;
    }

    // ...Internal JavaScript...
    if (!empty($internal_js_output)) {
      $comment = "<!-- Tidy Assets-> Consolidated Internal JavaScript";
      if ($use_minifier = getOption('tidy-assets_minify')) {
        if ($use_minifier == 'tidy-assets_jshrink') {
          $comment .= ", minified with JShrink";
          require_once 'tidy-assets/JShrink/Minifier.php';
          $internal_js_output = \JShrink\Minifier::minify($internal_js_output) . "\n";
        } else {
          $comment .= ", minified with JSqueeze";
          require_once 'tidy-assets/jsqueeze/JSqueeze.php';
          $jz = new \Patchwork\JSqueeze();
          $internal_js_output = $jz->squeeze($internal_js_output) . "\n";
        }
      }
      if (isset($start)) {
        echo $comment . " -->\n";
      }
      echo "<script>\n" . $internal_js_output . "</script>\n";
    }

    // ...Ending comment and processing time.
    if (isset($start)) {
      $end = microtime(true);
      echo "<!-- Tidy Assets End-> No changes after this line. Completed in "
      . round(($end - $start) * 1000, 2) . " ms -->\n";
    }
    /* NOTE! Uncomment below to print the filtering array for debugging */
    // echo "<!-- Filter's search strings\n======================\n";
    // print_r($skip);
    // echo "====================== -->\n";
  }
  /**
   * Stores matches of regex and remove them from searched string (if not on the filter list)
   * @author sbillard
   * @author bic-ed
   * @param  string $data     Before the first run, this is the collected HTML output (ob_get_contents()).
   *                          After that, already processed items are purged time by time.
   * @param  string $pattern  query string
   * @param  array  $filter   Search strings for detecting items not to be moved (bic-ed)
   * @return array            Filtered matches
   */
  static function extract(&$data, $pattern, $filter) {
    preg_match_all($pattern, $data, $matches);
    foreach ($matches[0] as $key => $found) {
      if (!empty($filter)) {
        if (tidyAssets::strposArray($matches[1][$key], $filter)) {
          unset($matches[1][$key]);
        } elseif (!empty($matches[1][$key])) { // Otherwise, parsing internal JS purges unmoved JS files
          $data = trim(str_replace($found, '', $data));
        }
      } else {
        $data = trim(str_replace($found, '', $data));
      }
    }
    return $matches;
  }
  /**
   * strpos on each value of an array
   * @param  string   $haystack The string to search in
   * @param  array    $needle   Each value is a search string
   * @return boolean            True on first match; False if no matches
   */
  static function strposArray($haystack, $needle) {
    foreach($needle as $what) {
        if((strpos($haystack, $what)) !== false) return true;
    }
    return false;
  }
}
